Add blog for JS abort controller (#1226)
Azure/azure-sdk: _posts/2020-04-21-js-abort-controller.md

---
title: Cancel operations in the Azure SDK with abort signals
layout: post
date: 21 apr 2020
sidebar: releases_sidebar
author_github: bterlson
repository: azure/azure-sdk
---

> Cancelling in-progress network operations is critical for many applications to maintain responsiveness and avoid waiting for pointless work to complete. For example, when downloading a large blob from Azure Storage, a user might want to cancel the download, and it would be nice if we could tell the Storage library that it doesn't need to download anymore. The new Azure SDK libraries for JavaScript and TypeScript have adopted abort signals for just this purpose. If you've used cancellation tokens in .NET, this should be very familiar to you.

Reviewer: anymore => any more

Author: I was going for the adverb form here, modifying the download verb, as opposed to saying we're not going to download any more [data]. I like the adverb form better but I'll change it if you insist 😀

> Cancelling in-progress network operations is critical for many applications to maintain responsiveness and avoid waiting for pointless work to complete. For example, when downloading a large blob from Azure Storage, a user might want to cancel the download, and it would be nice if we could tell the Storage library that it doesn't need to download anymore. The new Azure SDK libraries for JavaScript and TypeScript have adopted abort signals for just this purpose. If you've used cancellation tokens in .NET, this should be very familiar to you.

Reviewer: I think the last line reinforces we are .NET first.  I'd recommend removing.

Author: Will do.

## AbortController &amp; AbortSignal

`AbortController`s and `AbortSignal`s are the two types you use to cancel ongoing work. The controller is responsible for triggering cancellation, and signals are responsible for notifying when cancellation is requested. This separation of concerns enables you to safely pass an abort signal without delegating the ability to cancel the signal as well.

> `AbortController` and `AbortSignal` are standard features in the browser and are used with the `fetch` API. If you're just targeting the browser, you can use these built-in types and everything will work fine. If you're targeting Node, or if you want to take advantage of linked signals or other features that I'll cover later in this post, you can use the implementation in the Azure SDK found in the `@azure/abort-controller` npm package.

Reviewer: I think linking to MDN articles would be appropriate here.  Also, switching it round to be "Browsers already have a mechanism for aborting connections - here is how it works in general.  We've adopted it.", then going into "Of course, many services aren't run in the browser, so we've introduced the same classes as a polyfill for Node.  You can use these by including the @azure/abort-controller package."  with a link.

Then go onto "Using AbortController and AbortSignal with the Azure SDK" - you're basically introducing the idea, saying "look we work the same way" and then showing it off.

Author: These are good ideas. I'll add an MDN link and also swap the graf order.

const abortSignal = controller.signal;

// pass the abortSignal into the API you want to cancel
await blobClient.download({ abortSignal })

// then sometime later, cancel it!
controller.abort();
```

## Aborting multiple operations

Often you have multiple in-flight operations you might want to abort all at once. To continue our Storage example, maybe you're downloading multiple files in parallel. Or, perhaps you have to fetch a secret from Azure Keyvault before using that key in a subsequent operation. In that case, you can pass the same signal to each method, and calling abort on the controller will cancel each of them at the same time.

```javascript
async function getShoppingList({ abortSignal }) {
    const cs = await secretClient.getSecret('storage-connection-string', { abortSignal });
    const blobClient = new BlockBlobClient(cs.value, "MyDocuments", "shopping.rtf");
    return blobClient.downloadToBuffer(undefined,undefined, { abortSignal });
}

const controller = new AbortController();
const abortSignal = controller.signal;
const list = await getShoppingList({ abortSignal });

// then sometime later, cancel everything
controller.abort();
```

> In this example, we'll cancel fetching the connection string from Keyvault or, if we've already fetched that, cancel the download.

Reviewer: Future tense is bad.  we'll cancel => we cancel.

---
title: Cancel operations in the Azure SDK with abort signals
layout: post
date: 21 apr 2020
sidebar: releases_sidebar
author_github: bterlson
repository: azure/azure-sdk
---

> Cancelling in-progress network operations is critical for many applications to maintain responsiveness and avoid waiting for pointless work to complete. For example, when downloading a large blob from Azure Storage, a user might want to cancel the download, and it would be nice if we could tell the Storage library that it doesn't need to download any more. Or, maybe your process got an interrupt signal because the node is no longer needed or needs to upgrade and you need to stop any in-progress operations and shut down as soon as possible.

Reviewer 2: Consider:
For example, here are some scenarios where you could utilize the request cancelling functionality that is built into the Azure SDK.

- User started to download a large file from Azure Storage, but cancelled the request.
- A process got an interrupt signal because the node is no longer needed or needs to be upgraded.

Author: If you feel strongly I can change it but I prefer what's there personally.

Reviewer 2: If you like the paragraph form, consider rewording the last couple of sentences.

Here's an example:

Cancelling in-progress network operations is critical for many applications to maintain responsiveness and avoid waiting for pointless work to complete. For example, when downloading a large blob from Azure Storage, a user might want to cancel the download. It would be nice if the app could let the Storage library know that it can stop downloading it. Or, maybe your process got an interrupt signal because the node is no longer needed or needs to be upgraded. It would be nice if the app could stop any in-progress operations and shut down as soon as possible.

Author: Is your goal here to reduce sentence length/complexity? I will simplify. "it would be nice if" is a little wordy too, I'll try to get rid of it.

The new Azure SDK libraries for JavaScript and TypeScript have adopted abort signals for just these purposes.

## AbortController &amp; AbortSignal

> `AbortController` and `AbortSignal` are [standard features in the browser](https://developer.mozilla.org/en-US/docs/Web/API/AbortController) and are used with the [`fetch` API](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch) to abort in-progress network requests. The controller is responsible for triggering cancellation, and signals are responsible for notifying when cancellation is requested. This separation of concerns enables you to safely pass an abort signal without delegating the ability to cancel the signal as well.

Reviewer 3: > standard features in the browser

The page that this link goes to has a disclaimer that this is an experimental technology. Any concerns with calling an experimental tech to be standard? Or maybe add the link later so that we donta call something standard and experimental in the same line?

Author: It is standard in the sense that it is in the html dom specification and implemented by all major browsers. Going by MDN's [definition of experimental](https://developer.mozilla.org/en-US/docs/MDN/Contribute/Guidelines/Conventions_definitions#Experimental), they must think that the feature has risk of changing significantly, but I find this risk to be very low. Regardless, it *is* a standard 😀

> `AbortController` and `AbortSignal` are [standard features in the browser](https://developer.mozilla.org/en-US/docs/Web/API/AbortController) and are used with the [`fetch` API](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch) to abort in-progress network requests. The controller is responsible for triggering cancellation, and signals are responsible for notifying when cancellation is requested. This separation of concerns enables you to safely pass an abort signal without delegating the ability to cancel the signal as well.

Reviewer 3: > The controller is responsible for triggering cancellation, and signals are responsible for notifying when cancellation is requested.

Pick between "trigger" and "request" and use the same term in both places.

> `AbortController` and `AbortSignal` are [standard features in the browser](https://developer.mozilla.org/en-US/docs/Web/API/AbortController) and are used with the [`fetch` API](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch) to abort in-progress network requests. The controller is responsible for triggering cancellation, and signals are responsible for notifying when cancellation is requested. This separation of concerns enables you to safely pass an abort signal without delegating the ability to cancel the signal as well.

Reviewer 2: consider "tirggering the cancellation" and "when a cancellation"

Author: 👍

> `AbortController` and `AbortSignal` are [standard features in the browser](https://developer.mozilla.org/en-US/docs/Web/API/AbortController) and are used with the [`fetch` API](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch) to abort in-progress network requests. The controller is responsible for triggering cancellation, and signals are responsible for notifying when cancellation is requested. This separation of concerns enables you to safely pass an abort signal without delegating the ability to cancel the signal as well.

Reviewer 2: It's unclear what you mean by delegating the ability to cancel the signal as well.

Author: Holding a signal doesn't allow you to cancel it, only holding the controller does. This separation of concerns means you can safely pass around a signal and still have confidence things won't cancel because a library you passed it in to is misbehaving, for example.

Reviewer 2: Adding the text from your comment above would help.

Author: It's kind of a useless point I think. Objections to just deleting the last sentence "This separation of concerns enables you to safely pass an abort signal without delegating the ability to cancel the signal as well."?

Reviewer 2: Added comment to new PR.

> If you're just targeting the browser, you can use these built-in types and everything will work fine. If you're targeting Node, or if you want to take advantage of linked signals or other features that I'll cover later in this post, you can use the implementation in the Azure SDK found in the `@azure/abort-controller` npm package.

Reviewer 3: > you can use these built-in types

As a reader, I have no idea which built-in types you are referring to unless I clicked on the link in the previous paragraph. I would suggest we skip linking to https://developer.mozilla.org/en-US/docs/Web/API/AbortController in the previous para and do it here instead like so:

If you're just targeting the browser, you can use [browser built-in types for the AbortController](https://developer.mozilla.org/en-US/docs/Web/API/AbortController)

> If you're just targeting the browser, you can use these built-in types and everything will work fine. If you're targeting Node, or if you want to take advantage of linked signals or other features that I'll cover later in this post, you can use the implementation in the Azure SDK found in the `@azure/abort-controller` npm package.

Reviewer 3: > If you're targeting Node

I thought we had a deal to refer to it as Node.js!!

Author: I recall no such deal! But `Node` is wrong, I've standardized on `node` across the board.

Reviewer 3: But then you have 

> Or, maybe your process got an interrupt signal because the node is no longer needed 

Here node is a different node. Can we please please refer to node.js?

Author: Fiiiiiine

> If you're just targeting the browser, you can use these built-in types and everything will work fine. If you're targeting Node, or if you want to take advantage of linked signals or other features that I'll cover later in this post, you can use the implementation in the Azure SDK found in the `@azure/abort-controller` npm package.

Reviewer 3: > take advantage of linked signals

> found in the `@azure/abort-controller` npm package

link where you can my friend! Or atleast the first time a linkable entity is referenced. Candidates here are "linked signals" and the "@azure/abort-controller"

> If you're just targeting fairly modern browsers, you can use the built-in `AbortController` and `AbortSignal` and everything will work fine. If you're targeting Node, or if you want to take advantage of linked signals or other features that I'll cover later in this post, you can use the implementation in the Azure SDK found in the `@azure/abort-controller` npm package.

Reviewer 2: Consider "only" instead of "just"

Author: 👍

> If you're just targeting fairly modern browsers, you can use the built-in `AbortController` and `AbortSignal` and everything will work fine. If you're targeting Node, or if you want to take advantage of linked signals or other features that I'll cover later in this post, you can use the implementation in the Azure SDK found in the `@azure/abort-controller` npm package.

Reviewer 2: Please use Node.js

Author: It's Node.js in latest

> If you're just targeting fairly modern browsers, you can use the built-in `AbortController` and `AbortSignal` and everything will work fine. If you're targeting Node, or if you want to take advantage of linked signals or other features that I'll cover later in this post, you can use the implementation in the Azure SDK found in the `@azure/abort-controller` npm package.

Reviewer 2: Link to package.

> To abort an in-progress request such as an upload or download in storage, create a new `AbortController` and pass its signal into the method you might want to cancel later:

Reviewer 3: Do we want to call out anywhere that the abort signal doesnt really work on the service side? I don't want people to assume that they can cancel say a "create" operation and we will guarantee that the entity did not get created

Author: Yes this is a very good thing to point out.

> To abort an in-progress request such as an upload or download in storage, create a new `AbortController` and pass its signal into the method you might want to cancel later:

Reviewer 3: > To abort an in-progress

We are using a mix of "cancel" and "abort" in this doc. Can we stick to "cancel" when we need to use it as a verb and use "abort" only when referring to the package, the AbortError, the controller class, the method on the controller or the signal?

```javascript

> // If you want to use the AbortSignal built into the browser, omit this line

Reviewer 3: ```suggestion
// Omit this line if you want to use the AbortSignal built into the browser.
```

I always felt better about reading the action first and the reason later :)

> // If you want to use the AbortSignal built into the browser, omit this line

Reviewer 2: Just delete that line and the rest of the code with work?    How would I know to make this decision to use built-in or not?

Author: It's discussed in the earlier paragraphs - if you're targeting only browsers, and don't want to use linked signals or timeouts.

Author: And yes, you just delete the line and the rest will work!

import { AbortController } from "@azure/abort-controller";

// create a controller and get its signal
const controller = new AbortController();
const abortSignal = controller.signal;

// pass the abortSignal into the API you want to cancel
await blobClient.download({ abortSignal })

// then sometime later, cancel it!
controller.abort();
```

## Aborting multiple operations

> Often you have multiple in-flight operations you might want to abort all at once. To continue our Storage example, maybe you're downloading multiple files in parallel. Or, perhaps you have to fetch a secret from Azure KeyVault before using that key in a subsequent operation. In that case, you can pass the same signal to each method, and calling abort on the controller will cancel whichever of them is in progress at that time.

Reviewer 2: "operations that you"

> Often you have multiple in-flight operations you might want to abort all at once. To continue our Storage example, maybe you're downloading multiple files in parallel. Or, perhaps you have to fetch a secret from Azure KeyVault before using that key in a subsequent operation. In that case, you can pass the same signal to each method, and calling abort on the controller will cancel whichever of them is in progress at that time.

Reviewer 2: Key Vault with a space

> Often you have multiple in-flight operations you might want to abort all at once. To continue our Storage example, maybe you're downloading multiple files in parallel. Or, perhaps you have to fetch a secret from Azure KeyVault before using that key in a subsequent operation. In that case, you can pass the same signal to each method, and calling abort on the controller will cancel whichever of them is in progress at that time.

Reviewer 2: I wouldn't use "secret" here because that is official term in Key Vault, and it's confusing to say fetch a secret and then using a key.

I would change secret to key.

Author: I think key -> secret is more what I mean 😁

```javascript
async function getShoppingList({ abortSignal }) {
    const cs = await secretClient.getSecret('storage-connection-string', { abortSignal });
    const blobClient = new BlockBlobClient(cs.value, "MyDocuments", "shopping.rtf");
    return blobClient.downloadToBuffer(undefined,undefined, { abortSignal });
}

const controller = new AbortController();
const abortSignal = controller.signal;
const list = await getShoppingList({ abortSignal });

// then sometime later, cancel everything
controller.abort();
```

> In this example, we cancel fetching the connection string from KeyVault or, if we've already fetched that, cancel the download.

Reviewer 2: Key Vault

## Handling AbortErrors

> Looking at the above example you might wonder: if we abort while we're fetching the connection string from KeyVault, how do we avoid trying to create the `BlockBlobClient` with an empty key? If you've used AbortSignals with `fetch` in the browser, you know the answer already: cancelled operations throw an `AbortError`. In the above examples, that error is not handled - calling abort on a controller will trigger an unhandled exception which will get logged to the console in the browser and, sadly, cause Node to exit. We can fix this by handling the AbortError thrown from `getShoppingList`. Here's how we can add that functionality using our last example:

Reviewer 3: Am not very sure about the use of : here ...

Reviewer: It's correct usage.  However, I find the consecutive "if" statements jarring.  I think just asking the question is good enough:  "How do we avoid using an empty key when creating the `BlockBlobClient`?  Cancelled operations throw an `AbortError` when triggered.  In the above example, the `AbortError` is not handled, so the application will exit.  You should always handle `AbortError` gracefully.  For example:"

> Looking at the above example you might wonder: if we abort while we're fetching the connection string from KeyVault, how do we avoid trying to create the `BlockBlobClient` with an empty key? If you've used AbortSignals with `fetch` in the browser, you know the answer already: cancelled operations throw an `AbortError`. In the above examples, that error is not handled - calling abort on a controller will trigger an unhandled exception which will get logged to the console in the browser and, sadly, cause Node to exit. We can fix this by handling the AbortError thrown from `getShoppingList`. Here's how we can add that functionality using our last example:

Reviewer 2: Key Vault

> Looking at the above example you might wonder: if we abort while we're fetching the connection string from KeyVault, how do we avoid trying to create the `BlockBlobClient` with an empty key? If you've used AbortSignals with `fetch` in the browser, you know the answer already: cancelled operations throw an `AbortError`. In the above examples, that error is not handled - calling abort on a controller will trigger an unhandled exception which will get logged to the console in the browser and, sadly, cause Node to exit. We can fix this by handling the AbortError thrown from `getShoppingList`. Here's how we can add that functionality using our last example:

Reviewer 2: Node.js

```javascript
async function getShoppingList({ abortSignal }) {
    const cs = await secretClient.getSecret('storage-connection-string', { abortSignal });
    const blobClient = new BlockBlobClient(cs.value, "MyDocuments", "shopping.rtf");
    return blobClient.downloadToBuffer(undefined,undefined, { abortSignal });
}

const controller = new AbortController();
const abortSignal = controller.signal;

try {
    const list = await getShoppingList({ abortSignal });
} catch (e) {
    if (e.name === 'AbortError') {
        // abort was called on our abortSignal
        console.log('Operation was aborted by the user');
    } else {
        // some other error occurred 🤷‍♂️
        console.log('Downloading the shopping list failed');
    }
}

// then sometime later, cancel everything
controller.abort();
```

> ## Advanced tips and tricks

Reviewer 3: I would go with either `Tips and tricks" or "Advanced uses".

> The Azure SDK's implementation of AbortController has a few more tricks up its sleeve to make some very common and useful patterns easier. In order to take advantage of the following examples, you will need to use our `@azure/abort-controller` package - these APIs are not available built-in to the browser.

Reviewer 3: The first sentence already sets up the stage with tricks in the Azure SDK's implementation. The second one re-iterates the same.

```suggestion
The Azure SDK's implementation of AbortController via the `@azure/abort-controller` package has a few more tricks up its sleeve to make some very common and useful patterns easier as shown below. Please note that these APIs are not available in the built-in AbortController type in the browser.
```

### Aborting with a timeout

> Frequently you want to give an operation some amount of time to return a useful result, and if it hasn't, you're no longer interested. You could call `abortSignal.abort()` in a `setTimeout` callback, but this pattern is so common we provide a bit of sugar.

Reviewer 3: ```suggestion
Frequently you want to give an operation some amount of time to return a useful result, and if it hasn't, you're no longer interested. You could call `abortSignal.abort()` in a `setTimeout` callback, but this pattern is so common that we decided to provide a bit of sugar.
```

> Frequently you want to give an operation some amount of time to return a useful result, and if it hasn't, you're no longer interested. You could call `abortSignal.abort()` in a `setTimeout` callback, but this pattern is so common we provide a bit of sugar.

Reviewer 2: Consider rewording "you're no longer interested" to "you'd like to cancel it."

Author: 👍

```javascript
const list = await getShoppingList({ abortSignal: AbortController.timeout(1000) });

// roughly equivalent to:
const controller = new AbortController();
const abortSignal = controller.signal;
const list = await getShoppingList({ abortSignal });

> setTimeout(() => controller.abort(), 1000);

Reviewer 3: Can we split the above into two different code snippets to highlight the difference and the sweetness of the sugar you have provided :) ?

```javascript
const list = await getShoppingList({ abortSignal: AbortController.timeout(1000) });
```

This is roughly equivalent to the below:
```javascript
// rest goes here
```

```

### Linked signals

> In many real-world applications, in-progress operations often need to get cancelled for a variety of reasons. For example, you might want to cancel a download after a certain amount of time has elapsed or the user presses the cancel button. The Azure SDK's `AbortController` supports linked signals for this purpose - when a signal is aborted, any of its linked signals are also aborted. To show how you might use this in practice, we'll modify our previous example some more by making `getShoppingList` responsible for aborting operations that are taking too long:

Reviewer 3: > In many real-world applications, in-progress operations often need to get cancelled for a variety of reasons.

For a while, I kept wondering that this statement isnt telling me anything that I havent already read in this document. Only after going to the end I realized that we want to highlight the case of cancelling due to more than 1 reason. This is because `variety of abc` is such a generic term.

Can we change `variety of reasons` to `more than one reason` and say something like `you might wonder that you will now need to create a separate abort signal for each reason and pass each one of them to an operation`

Reviewer: Avoid "you might wonder" - it never reads well.

---

> title: Cancel operations in the Azure SDK with abort signals

Reviewer 3: Should the title reflect that the blog is specific to JS/TS?

Author: I could put "Azure SDK for JavaScript and TypeScript" but it makes it so long....

Reviewer: HOw about just "How to cancel operations in the Azure SDK for JavaScript" - I know the TypeScript isn't there, but it will get tagged as well.

Author: Going with Adrian's suggestion.

> title: Cancel operations in the Azure SDK with abort signals

Reviewer 2: Recommend this title:
How to use abort signals to cancel operations in the Azure SDK for JavaScript/TypeScript

Author: It's very long, but ok!

layout: post
date: 21 apr 2020
sidebar: releases_sidebar
author_github: bterlson
repository: azure/azure-sdk
---

Cancelling in-progress network operations is critical for many applications to maintain responsiveness and avoid waiting for pointless work to complete. For example, when downloading a large blob from Azure Storage, a user might want to cancel the download, and it would be nice if we could tell the Storage library that it doesn't need to download any more. Or, maybe your process got an interrupt signal because the server is no longer needed or needs to upgrade and you need to stop any in-progress operations and shut down as soon as possible.

> The new Azure SDK libraries for JavaScript and TypeScript have adopted abort signals for just these purposes.

Reviewer: Remove "new"

        // abort was called on our abortSignal
        console.log('Operation was aborted by the user');
    } else {
        // some other error occurred 🤷‍♂️
        console.log('Downloading the shopping list failed');
    }
}

// then sometime later, cancel everything
controller.abort();
```

## Advanced uses

`@azure/abort-controller`'s implementation of `AbortController` has a couple more tricks up its sleeve to make some common patterns easier. Please note that these APIs are not available in the browser's `AbortController`.

### Cancelling with a timeout

Frequently you want to give an operation some amount of time to return a useful result, and if it hasn't, you're no longer interested. You could call `abortSignal.abort()` in a `setTimeout` callback, but this pattern is so common we provide a bit of sugar:

```javascript
const list = await getShoppingList({ abortSignal: AbortController.timeout(1000) });
```

This is roughly equivalent to the following code:

```javascript
const controller = new AbortController();
const abortSignal = controller.signal;
const list = await getShoppingList({ abortSignal });
setTimeout(() => controller.abort(), 1000);
```

### Linked signals

> In many real-world applications, in-progress operations often need to get cancelled for multiple reasons. For example, you might want to cancel a download after a certain amount of time has elapsed and when the user presses the cancel button. The Azure SDK's `AbortController` supports linked signals for this purpose - when a signal is aborted, any of its linked signals are also aborted.

Reviewer 3: ```suggestion
In many real-world applications, in-progress operations often need to get cancelled for multiple reasons. For example, you might want to cancel a download after a certain amount of time has elapsed or when the user presses the cancel button. The Azure SDK's `AbortController` supports linked signals for this purpose - when a signal is aborted, any of its linked signals are also aborted.
```

    const cs = await secretClient.getSecret('storage-connection-string', {
        // pass our fancy linked signal
        abortSignal: controller.signal
    });
    const blobClient = new BlockBlobClient(cs.value, "MyDocuments", "shopping.rtf");
    return blobClient.downloadToBuffer(undefined,undefined, {
        abortSignal: controller.signal
    });
}

const controller = new AbortController();
const abortSignal = controller.signal;
const list = await getShoppingList({ abortSignal });

// if user indicates they're done waiting
controller.abort();
```

Linked signals are also useful when an operation is divided into many sub-tasks and you want granular control over which parts of the task you cancel while also being able to cancel the entire operation all at once. The top-level task might even be the application itself. For example, if you want to handle an interrupt signal gracefully by cancelling all in-progress operations inside your application, you can create an abort controller at application startup, handle the interrupt signal by calling `abort()`, and pass that signal or linked signals to all calls into Azure SDK libraries.

## Further reading

> * [@azure/abort-controller readme on npm](https://www.npmjs.com/package/@azure/abort-controller)

Reviewer: Should readme be README (caps)?

Author: That's just a file naming convention, no need to scream it IMO 😁
